crates/redis/src/lib.rs

```diff
@@ -52,30 +54,91 @@ impl TriggerExecutor for RedisTrigger {
     type RunConfig = NoArgs;
 
     async fn new(engine: TriggerAppEngine<Self>) -> Result<Self> {
-        let address = engine
+        let default_address: String = engine
             .trigger_metadata::<TriggerMetadata>()?
             .unwrap_or_default()
             .address;
 
-        let mut channel_components: HashMap<String, Vec<String>> = HashMap::new();
+        let mut server_channels: HashMap<String, ChannelComponents> = HashMap::new();
 
         for (_, config) in engine.trigger_configs() {
-            channel_components
+            let address = config.address.clone().unwrap_or(default_address.clone());
+            let server = server_channels.entry(address).or_default();
+            server
                 .entry(config.channel.clone())
                 .or_default()
                 .push(config.component.clone());
         }
         Ok(Self {
-            engine,
-            address,
-            channel_components,
+            engine: Arc::new(engine),
+            server_channels,
         })
     }
 
     /// Run the Redis trigger indefinitely.
     async fn run(self, _config: Self::RunConfig) -> Result<()> {
-        let address = &self.address;
+        let addresses: Vec<String> = self
+            .server_channels
+            .keys()
+            .map(|key| key.to_owned())
+            .collect();
+        let tasks: Vec<_> = addresses
+            .into_iter()
+            .map(|address| {
+                let trigger = self.clone();
+                // safe to unwrap hashmap as only address that exist are passed
```

Comment:
Rather than doing this, consider iterating the HashMap, which will give you a tuple of the address with its channel map.  No need to extract the address then later go back and look up the channel map.  A la

```
let tasks = self.server_channels.into_iter().map(|server, channel_components| ...)
```